### PR TITLE
Added userns support on docker_container module Fixes #20648

### LINKED ISSUE #20648: Feature idea: docker_container support for --userns
_From @ko-christ on September 27, 2016 13:46_

##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

docker_container
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

Docker allows setting the `userns` mapping, but docker_container currently does not support this functionality. It would be great to add support for this in `docker_container` module as well.

`userns_mode` also seems to be available with `docker-py` version `1.10.0`.
https://github.com/docker/docker-py/pull/1113

_Copied from original issue: ansible/ansible-modules-core#5054_

### PR DESCRIPTION
**Pull request by hdimitriou**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
modules/cloud/docker/docker_container

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel_docker_container_userns f8fd45238f) last updated 2017/03/02 14:53:34 (GMT +300)
```

##### SUMMARY

Defining userns in the docker_container module is useful in certain scenarios. 
Currently docker-py module only supports userns_mode = host, so the most useful case for this is when you have defined userns on your docker daemon and you need to start a container with network_mode = host (which also requires userns_mode = host). 
Fixes #20648

```
Before: 
fatal: [127.0.0.1]: FAILED! => {"changed": false, "failed": true, "msg": "unsupported parameter for module: userns_mode"}

After:
changed: [127.0.0.1]
```

### CODE REVIEW

**Review by kassiansun — APPROVED** [review]
LGTM

**Comment by kassiansun** [comment]
BTW, we have many features like this, which is only supported by `docker >= 1.2x`, or supported by `docker-py >= 2.x`, we should find a good way to handle this, or we might get flooded by future issue reports.

**Comment by ningyougang** [comment]
@kassiansun , do you mean, if we merged this PR, because of  only supported by `docker >= 1.2X and docker-py>= 2.X`, it will have more issues in future?

**Comment by kassiansun** [comment]
@ningyougang People might use this feature without knowing the fact that it's supported by a specific version of `docker` or `docker-py`, and report it as an issue.

Actually I think everyone should upgrade to `docker` >= 1.12 and latest `docker-py`, `docker` is very unstable, especially `docker-py`. Sticking with latest version will make it more usable, as community always focus the most attention on latest package.

**Comment by ningyougang** [comment]
@kassiansun , as far as i know, docker for python lib has below libs
```
oot@lambda-nyg ~]# pip list|grep docker
docker (2.2.1)
docker-py (1.10.6)
```
it supports docker , or docker-py?

**Comment by kassiansun** [comment]
@ningyougang Both, but latest `docker` is personally recommended

**Comment by ThomasSteinbach** [comment]
Sorry @chouseknecht  @cove @dusdanig @jctanner @joshuaconner @kassiansun @softzilla @zfil

Does anybody from you know how to unsubscribe from the maintainers list?

**Comment by pilou-** [comment]
@ThomasSteinbach all authors are maintainers by default ([see](https://github.com/hdimitriou/ansible/blob/57d12db8040c8ace0fcfe11727f08c016fce9f92/lib/ansible/modules/cloud/docker/docker_container.py#L474)), but it seems you're already [excluded](https://github.com/ansible/ansible/blob/96b4f8b5c3d70055e0aa065757e415c9fe643872/.github/BOTMETA.yml#L219) from the maintainers list (since 7596826ed01b92eb69baff55609f15ec394435f3).

File: lib/ansible/modules/cloud/docker/docker_container.py [hunk 1]
```diff
@@ -266,6 +266,12 @@
       - none
     default: null
     required: false
+  userns_mode:
+     description:
+       - User namespace to use
+     default: null
+     required: false
+     version_added: "2.3"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/docker/docker_container.py]: `version_added: "2.5"`

**Review by chouseknecht — APPROVED** [review]

**Comment by chouseknecht** [comment]
shipit

**Review by gundalow — APPROVED** [review]

File: lib/ansible/modules/cloud/docker/docker_container.py (hunk 1 above)
Comment (gundalow) [reply lib/ansible/modules/cloud/docker/docker_container.py]: I've just fixed this directly in the PR

**Pull request merged**